### PR TITLE
Added Copy() function on vm.Env

### PR DESCRIPTION
**Pull request by c4s4**
I need this function to clone a VM in the [NeON build tool](http://github.com/c4s4/neon). As *vm.Env* fields are not public, I have to fork Anko project to implement this *Copy()* function. This would be great if this was integrated in the Anko project.

### CODE REVIEW

**Comment by MichaelS11** [comment]
@c4s4 
Would be great if tests could be added. If you would like details about adding tests, just let me know.

**Comment by c4s4** [comment]
@MichaelS11 
I have added tests for Copy() function.

File: vm/copy_test.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"testing"
+)
+
+func TestCopy(t *testing.T) {
+	env := NewEnv()
+	env.Define("a", "a")
+	copy := env.Copy()
+	if v, e := copy.Get("a"); e != nil || v != "a" {
+		t.Errorf("Copy doesn't retain original values")
```
Comment (mattn) [inline vm/copy_test.go]: please use lower case string in first word if it is not identifier. and below.

**Comment by MichaelS11** [comment]
I think the copy function could go into the env.go and env_test.go files.

File: vm/copy.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import "reflect"
+
+// Copy the state of the virtual machine environment
+func (env *Env) Copy() *Env {
+	b := false
+	copy := Env {
+		env:       make(map[string]reflect.Value),
+		typ:       make(map[string]reflect.Type),
+		parent:    nil,
```
Comment (MichaelS11) [inline vm/copy.go]: Should the parent be copied?

File: vm/copy_test.go (hunk 1 above)
Comment (c4s4) [reply vm/copy_test.go]: Done

File: vm/copy.go (hunk 2 above)
Comment (c4s4) [reply vm/copy.go]: Done

**Comment by mattn** [comment]
LGTM. pass to @MichaelS11 's look.

**Comment by mitchmieuxplacer** [comment]
Regarding recursive copy, we could add a recursive boolean argument telling if copy is recursive.

**Comment by mattn** [comment]
Or add new DeepCopy?

**Comment by mattn** [comment]
So is this Clone not Copy? How do you think?

**Comment by mitchmieuxplacer** [comment]
I can't run tests, they fail on master.

**Comment by mitchmieuxplacer** [comment]
I must say I don't really see a difference between clone and copy...

**Comment by MichaelS11** [comment]
What about name and external variables?

File: vm/env.go [hunk 3]
```diff
@@ -422,3 +422,42 @@ func (e *Env) Execute(src string) (interface{}, error) {
...
+		typ:       make(map[string]reflect.Type),
+		parent:    e.parent,
+		interrupt: &b,
+	}
+	for name, value := range e.env {
+		copy.env[name] = value
+	}
+	for name, typ := range e.typ {
+		copy.typ[name] = typ
+	}
```
Comment (MichaelS11) [inline vm/env.go]: Can you just call copy instead of the above code?

**Comment by MichaelS11** [comment]
What about name and external variables?

**Comment by c4s4** [comment]
Done

File: vm/env.go [hunk 4]
```diff
@@ -422,3 +422,32 @@ func (e *Env) Execute(src string) (interface{}, error) {
 func (e *Env) Run(stmts []ast.Stmt) (interface{}, error) {
 	return Run(stmts, e)
 }
+
+// Copy the state of the virtual machine environment
+func (e *Env) Copy() *Env {
+	b := false
+	copy := Env {
```
Comment (MichaelS11) [inline vm/env.go]: It doesn't really matter, but would be kind of nice if it was in the same order as the struct declare.
type Env struct {
	name      string
	env       map[string]reflect.Value
	typ       map[string]reflect.Type
	parent    *Env
	interrupt *bool
	external  EnvResolver

File: vm/env.go [hunk 5]
```diff
@@ -422,3 +422,32 @@ func (e *Env) Execute(src string) (interface{}, error) {
...
+		interrupt: &b,
+		name:      e.name,
+		external:  e.external,
+	}
+	for name, value := range e.env {
+		copy.env[name] = value
+	}
+	for name, typ := range e.typ {
+		copy.typ[name] = typ
+	}
```
Comment (MichaelS11) [inline vm/env.go]: Should there be a lock when coping the maps?

File: vm/env.go (hunk 4 above)
Comment (c4s4) [reply vm/env.go]: Reordered.

File: vm/env.go (hunk 5 above)
Comment (c4s4) [reply vm/env.go]: Added lock on copy.

**Comment by mattn** [comment]
Thank you

**Comment by mitchmieuxplacer** [comment]
Thanks

**Pull request merged**